commons-executors/src/main/java/com/palantir/common/concurrent/ExecutorInheritableThreadLocal.java

```diff
@@ -116,13 +139,13 @@ protected T initialValue() {
         return null;
     }
 
-    static WeakHashMap<ExecutorInheritableThreadLocal<?>, Object> getMapForNewThread() {
-        WeakHashMap<ExecutorInheritableThreadLocal<?>, Object> currentMap = mapForThisThread.get();
+    static ConcurrentMap<ExecutorInheritableThreadLocal<?>, Object> getMapForNewThread() {
+        ConcurrentMap<ExecutorInheritableThreadLocal<?>, Object> currentMap = perThreadThreadCache.get().asMap();
         if (currentMap.isEmpty()) {
-            mapForThisThread.remove();
+            perThreadThreadCache.remove();
             return currentMap;
         }
-        WeakHashMap<ExecutorInheritableThreadLocal<?>, Object> ret = new WeakHashMap<ExecutorInheritableThreadLocal<?>, Object>(currentMap.size());
+        ConcurrentMap<ExecutorInheritableThreadLocal<?>, Object> ret = new ConcurrentHashMap<ExecutorInheritableThreadLocal<?>, Object>(currentMap.size());
         for (Map.Entry<ExecutorInheritableThreadLocal<?>, Object> e : currentMap.entrySet()) {
             ret.put(e.getKey(), e.getKey().callChildValue(e.getValue()));
```

Comment:
ok, the other possibility is that callChildValue does sets into the map for the current thread.